src/main/java/roboy/dialog/states/ordinaryStates/FUAState.java

```diff
@@ -0,0 +1,103 @@
+import roboy.memory.nodes.MemoryNodeModel;
+import roboy.util.QAJsonParser;
+import roboy.util.RandomList;
+
+import java.util.List;
+import java.util.Set;
+
+import static roboy.memory.Neo4jRelationships.*;
+
+/**
+ * Follow Up Asking State
+ * This state is only entered if there are some known facts about the active interlocutor.
+ * It asks if the known facts are still up to date.
+ */
+public class FUAState extends State {
+
+    private QAJsonParser qaValues;
+    private Neo4jRelationships[] predicates = { FROM, HAS_HOBBY, WORK_FOR, STUDY_AT };
+    private Neo4jRelationships selectedPredicate;
+    private State nextState;
+
+    private final String TRANSITION_INFO_UPDATED = "questionAnswering";
+    private final Logger LOGGER = LogManager.getLogger();
+
+    public FUAState(String stateIdentifier, StateParameters params) {
+        super(stateIdentifier, params);
+        qaValues = new QAJsonParser(params.getParameter("qaFile"));
+    }
+
+    @Override
+    public State.Output act() {
+        Interlocutor person = Context.getInstance().ACTIVE_INTERLOCUTOR.getValue();
+
+        for (Neo4jRelationships predicate : predicates) {
+            if (person.hasRelationship(predicate) && !Context.getInstance().DIALOG_INTENTS.contains(new IntentValue("FUP", predicate))) {
+                selectedPredicate = predicate;
+                break;
+            }
+        }
+
+        if (selectedPredicate != null) {
+            RandomList<String> questions = new RandomList<>(qaValues.getFollowUpQuestions(selectedPredicate));
+            String retrievedResult = "";
+            RandomList<MemoryNodeModel> nodes = retrieveNodesFromMemoryByIds(person.getRelationships(selectedPredicate));
+            if (!nodes.isEmpty()) {
+                retrievedResult = nodes.getRandomElement().getProperties().get("name").toString();
+            }
+            String question = String.format(questions.getRandomElement(), retrievedResult);
+            Context.getInstance().DIALOG_INTENTS_UPDATER.updateValue(new IntentValue("FUP", selectedPredicate));
+            return Output.say(question);
+        } else {
+            return Output.sayNothing();
+        }
+    }
+
+    @Override
+    public State.Output react(Interpretation input) {
+        // TODO: How do we ensure it is the same Interlocutor in the same state of existence?
+        Interlocutor person = Context.getInstance().ACTIVE_INTERLOCUTOR.getValue();
+        String answer = "I have no words";
+
+        // TODO: Will need to consider proper conditions for processing
+        if (selectedPredicate != null) {
+            // TODO: Perform updating the person object
+            Context.getInstance().ACTIVE_INTERLOCUTOR_UPDATER.updateValue(person);
+            List<String> answers = qaValues.getFollowUpAnswers(selectedPredicate);
+            if (answers != null && !answers.isEmpty()) {
+                answer = String.format(answers.get((int)(Math.random() * answers.size())), "");
```

Comment:
Please change the return type of ``qaValues.getFollowUpAnswers`` from List to RandomList so we can get rid of all those ``Math.random()`` calls.

Reply:
Done